File: Engine/core/ipc.py
"""JSON-RPC IPC communication."""
import json
import sys


def send_response(obj: dict):
    try:
        sys.stdout.write(json.dumps(obj) + "\n")
        sys.stdout.flush()
    except (BrokenPipeError, OSError):
        pass


def send_progress(percent: int):
    send_response({"type": "progress", "percent": percent})


def send_finished(ok: bool, message: str = "", file_path: str = ""):
    send_response({"type": "finished", "ok": ok, "message": message, "file_path": file_path})


def send_download_status(percent=None, speed=None, eta=None, is_live=False, status=""):
    send_response({
        "type": "download_status",
        "percent": percent,
        "speed": speed,
        "eta": eta,
        "is_live": is_live,
        "status": status,
    })

File: cpp_engine/src/ipc.cpp
#include "ipc.h"
#include <iostream>
#include <nlohmann/json.hpp>

namespace engine {

Ipc& Ipc::instance() {
    static Ipc inst;
    return inst;
}

void Ipc::send_response(const std::string& json_str) {
    std::lock_guard<std::mutex> lock(write_mutex_);
    try {
        std::cout << json_str << "\n" << std::flush;
    } catch (...) {}
}

void Ipc::send_progress(int percent) {
    nlohmann::json j;
    j["type"] = "progress";
    j["percent"] = percent;
    send_response(j.dump());
}

void Ipc::send_finished(bool ok, const std::string& message, const std::string& file_path) {
    nlohmann::json j;
    j["type"] = "finished";
    j["ok"] = ok;
    j["message"] = message;
    j["file_path"] = file_path;
    send_response(j.dump());
}

}
